mypy/ipc.py

```diff
@@ -0,0 +1,185 @@
+                try:
+                    self.connection = _winapi.CreateFile(
+                        self.name,
+                        _winapi.GENERIC_READ | _winapi.GENERIC_WRITE,
+                        0,
+                        _winapi.NULL,
+                        _winapi.OPEN_EXISTING,
+                        0,
+                        _winapi.NULL,
+                    )
+                except WindowsError as e:
+                    if e.winerror == _winapi.ERROR_PIPE_BUSY:
+                        raise IPCException("The connection is busy.")
+                else:
+                    break
+            _winapi.SetNamedPipeHandleState(self.connection,
+                                            _winapi.PIPE_READMODE_MESSAGE,
+                                            None,
+                                            None)
+        else:
+            self.connection = socket.socket(socket.AF_UNIX)
+            self.connection.settimeout(timeout)
+            self.connection.connect(name)
+
+    def __enter__(self) -> 'IPCClient':
+        return self
+
+
+class IPCServer(IPCBase):
+
+    BUFFER_SIZE = 2**16
+
+    def __init__(self, name: str, timeout: Optional[int] = None) -> None:
+        super().__init__(name)
+        if sys.platform != 'win32':
+            self.sock_directory = tempfile.mkdtemp()
+            sockfile = os.path.join(self.sock_directory, self.name)
+            self.sock = socket.socket(socket.AF_UNIX)
+            self.sock.bind(sockfile)
+            self.sock.listen(1)
+            if timeout is not None:
+                self.sock.settimeout(timeout)
+
+    def __enter__(self) -> 'IPCServer':
+        if sys.platform == 'win32':
+            # The NamedPipe needs to be recreated every time we wait for a connection
+            self.connection = _winapi.CreateNamedPipe(self.name,
+                _winapi.PIPE_ACCESS_DUPLEX | _winapi.FILE_FLAG_FIRST_PIPE_INSTANCE,
+                _winapi.PIPE_READMODE_MESSAGE | _winapi.PIPE_TYPE_MESSAGE | _winapi.PIPE_WAIT,
+                1,  # one instance
+                self.BUFFER_SIZE,
+                self.BUFFER_SIZE,
+                0,  # PIPE_WAIT, wait for a connection of the client
+                0,  # PIPE_ACCEPT_REMOTE_CLIENTS (this is the default, so I guess?)
+                                                      )
+            if self.connection == -1:  # INVALID_HANDLE_VALUE
+                err = _winapi.GetLastError()
+                raise IPCException('Invalid handle to pipe: {err}'.format(err))
+            # NOTE: It is theoretically possible that this will hang forever if the
+            # client never connects, though this can be "solved" by killing the server
+            try:
+                _winapi.ConnectNamedPipe(self.connection, _winapi.NULL)
+            except WindowsError as e:
+                if e.winerror == _winapi.ERROR_PIPE_CONNECTED:
+                    pass  # The client already exists, which is fine.
```

Comment:
Do you mean to throw in all other cases? This does a check but ignores everything regardless.

Reply:
Thanks I fixed this and a case below.